src/common/runtime/src/lib.rs

```diff
@@ -81,36 +80,25 @@ impl Runtime {
         rx.recv().expect("Spawned task transmitter dropped")
     }
 
-    /// Spawn a task on the runtime and await on it.
-    /// You should use this when you are spawning compute or IO tasks from the Executor.
-    pub async fn await_on<F>(&self, future: F) -> DaftResult<F::Output>
-    where
-        F: Future + Send + 'static,
-        F::Output: Send + 'static,
-    {
-        let (tx, rx) = oneshot::channel();
-        let pool_type = self.pool_type;
-        let _join_handle = self.spawn(async move {
-            let task_output = Self::execute_task(future, pool_type).await;
-            if tx.send(task_output).is_err() {
-                log::warn!("Spawned task output ignored: receiver dropped");
-            }
-        });
-        rx.await.expect("Spawned task transmitter dropped")
-    }
-
     /// Blocks current thread to compute future. Can not be called in tokio runtime context
     ///
     pub fn block_on_current_thread<F: Future>(&self, future: F) -> F::Output {
         self.runtime.block_on(future)
     }
 
-    pub fn spawn<F>(&self, future: F) -> JoinHandle<F::Output>
+    // Spawn a task on the runtime
+    pub fn spawn<F>(
+        &self,
+        future: F,
+    ) -> impl Future<Output = Result<F::Output, JoinError>> + Send + 'static
     where
         F: Future + Send + 'static,
         F::Output: Send + 'static,
     {
-        self.runtime.spawn(future)
+        // Spawn it on a joinset on the runtime, such that if the future gets dropped, the task is cancelled
+        let mut joinset = tokio::task::JoinSet::new();
+        joinset.spawn_on(future, self.runtime.handle());
+        async move { joinset.join_next().await.expect("just spawned task") }.boxed()
```

Comment:
we shouldn't have to box this right? We should avoid the heap allocation if we can here

Reply:
I also think it would be cleaner to make a struct that holds the joinset and then impl the future trait on that struct.

Similarly to what influx does for the their `Job` struct
https://github.com/metrico/influxdb_iox/blob/ab17bbc9efbb8568ea5a95ccb9d4bbddd33fc9ea/executor/src/lib.rs#L88

Reply:
yeah good catch no box needed since there's only 1 branch

Reply:
Encapsulated the fut in a `RuntimeTask` struct